### DISCUSSION TITLE
View の更新トリガーを調べるために `_printChanges()` を利用する。

### DISCUSSION THREAD

**Discussion by YusukeHosonuma** [YusukeHosonuma/Effective-SwiftUI#24] [Ideas]
## Overview

SwiftUI 3.0 において、開発者向けのデバッグ機能として `_printChanges()` というメソッドが導入された。

これは View の更新をトリガーした値の変更などを調べることができるもので、`body` 関数の先頭で以下のように記述して利用する。

```swift
struct ContentView: View {
    @State var x = 1

    var body: some View {
        let _ = Self._printChanges() // ✅ Dump the information that triggered the View update.

        Button("x++") { x += 1 }
        Text("\(x)")
    }
}
```

上記を初期表示し、`x++`ボタンをタップすると、コンソールには以下の内容が出力される。

```
ContentView: @self, @identity, _x changed. // 💡 Initial representation.
ContentView: _x changed. // 💡 Button tapped.
```

このように対象の View が更新されたタイミングで、そのトリガーとなった値などが出力される。このケースでは `_x changed` と出力されており、 `@State` の内部値（`wrapped-value`）が更新契機となっていることが分かる（`@self` や `@identity` はこのあと説明する）。

これは意図せず大量に更新が走っている際に、その原因を突き止めるのに役立つはずだ。

この API は（頭に `_` がついていることからも察せられるとおり）プライベートAPI という位置づけになっている。そのため、Apple からのドキュメントは提供されていないが、Xcode 上からヘルプを表示することで確認できる。

<img width="535" alt="image" src="https://user-images.githubusercontent.com/2990285/162358938-63a7ecd9-9c7d-4ae4-8a08-943450d386b9.png">

> When called within an invocation of body of a view of this type, prints the names of the changed dynamic properties that caused the result of body to need to be refreshed. As well as the physical property names, “@self” is used to mark that the view value itself has changed, and “@identity” to mark that the identity of the view has changed (i.e. that the persistent data associated with the view has been recycled for a new instance of the same type).

この説明内容から `@self` や `@identity` が意味するところも分かる。

### `@self`

`@self` は View 自体が更新されたことを意味する。

```swift
struct ContentView: View {
    @State var x = 1

    var body: some View {
        Button("x++") { x += 1 }
        NumberView(count: x)
    }
}

struct NumberView: View {
    var count: Int
    
    var body: some View {
        let _ = Self._printChanges() // 💡
        Text("count: \(count)")
    }
}
```

上記を初期表示してから `x++` ボタンをタップすると、以下のように出力される。

```
NumberView: @self changed.
NumberView: @self changed.
```

これは`ContentView`の`x`が更新され、それによって`NumberView`が再生成（`@self changed`）されたことを表す。

### `@identity`

`@identity` は View の（差分計算のための）ID 自体が変化したことを意味する。

```swift
struct ContentView: View {
    @State var viewID = 1

    var body: some View {
        Button("viewID++") { viewID += 1 }
        XView()
            .id(viewID) // 💡 Specify id by itself.
    }
}

struct XView: View {
    @State var x = 0

    var body: some View {
        let _ = Self._printChanges()
        Text("x: \(x)")
    }
}
```

上記を初期表示してから`viewID++` ボタンをタップすると、以下のように出力される。

```
CounterView: @self, @identity, _x changed.
CounterView: @self, @identity, _x changed.
```

他の値も出力されていて分かりづらいが、初期表示によって ID が振られたことによって `@identity` が出力され、その後の ID 変更によって再び `@identity` が出力されたことが読み取れるはずだ。

ここで `XView` にわざわざ `@State var x = 0` を用意していることに留意して欲しい。これはその View に保持する状態がなければ、SwiftUI はそもそも View の変更を検知する必要がないため、差分計算のために `XView` を View ツリーに含める必要がなくなるためだ（おそらく）。

そのため `@State var x = 0`を取り除くと、`@self changed.` としか出力されなくなる。

## Discussion

ここまで解説した`@self`および`@identity`については、私の憶測も多分に含まれる。おそらく、この理解で正しいと思うが、間違いがありそうであれば指摘して欲しい。

## Why not?

プライベート API という位置づけであるため、このコードを含めた状態でアプリを提出するとリジェクトされる可能性があるので注意すること。

## Reference

- [id()](https://developer.apple.com/documentation/swiftui/ellipticalgradient/id(_:))

**Comment by externvoid** [comment]
idメソッドで値をセットするとViewの識別子が変化するとReferenceにはあります。Viewの識別子(the identity of the view)はなにか？をNew Bingに尋ねるとView階層内での一意な値で変更されると新しいViewだとみなされるのだそうです。Viewの再描画は
1. 状態変数の変化
2. View階層の変化

によって起こるので、idメソッドは強制再描画を起こすために使えそうです。また、＠identityの表記はView階層に変化により再描画が行われた事を示しているようです。
